File: crop-me/src/main/java/com/mikelau/croperino/CroperinoFileUtil.java
package com.mikelau.croperino;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.MediaStore;
import android.support.v4.app.ActivityCompat;

import com.mikelau.magictoast.MagicToast;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.channels.FileChannel;

/**
 * Created by Mike on 9/15/2016.
 */
public class CroperinoFileUtil {

    public static final int REQUEST_EXTERNAL_STORAGE = 1;
    public static final int REQUEST_CAMERA = 2;
    public static String[] PERMISSIONS_STORAGE = {
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    };

    public static File mFileTemp;

    public static File getmFileTemp() {
        return mFileTemp;
    }

    public static void setupDirectory(Context ctx) {
        String state = Environment.getExternalStorageState();
        if (Environment.MEDIA_MOUNTED.equals(state)) {
            if (Environment.getExternalStorageDirectory().exists()) {
                mFileTemp = new File(Environment.getExternalStorageDirectory() + CroperinoConfig.getsDirectory(), CroperinoConfig.getsImageName());
            } else {
                mFileTemp = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM) + CroperinoConfig.getsDirectory(), CroperinoConfig.getsImageName());
            }
        } else {
            mFileTemp = new File(ctx.getFilesDir(), CroperinoConfig.getsImageName());
        }

        if (!mFileTemp.exists()) {
            mFileTemp = new File(CroperinoConfig.getsRawDirectory());
            mFileTemp.mkdirs();
        }
    }

    public static File newCameraFile() throws IOException {
        File storageDir = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES), "Camera");
        if(!storageDir.exists()) {
            storageDir.mkdir();
        }
        mFileTemp = File.createTempFile(CroperinoConfig.getsImageName(), ".jpg", storageDir);
        return mFileTemp;
    }

    public static File newGalleryFile(Intent data, Context ctx) {
        try {
            mFileTemp = new File(CroperinoConfig.getsRawDirectory() + CroperinoConfig.getsImageName());
            copyFile(new File(CroperinoFileUtil.getPath(ctx, data.getData())), mFileTemp);
            return mFileTemp;
        } catch (Exception e) {
            if(e instanceof  IOException) {
                mFileTemp = new File(CroperinoFileUtil.getPath(ctx, data.getData()));
            } else {
                MagicToast.showError(ctx, "Gallery is empty or access is prohibited by device.");
            }
            return mFileTemp;
        }

    }

    private static void copyFile(File sourceFile, File destFile) throws IOException {
        if (!sourceFile.exists()) {
            return;
        }

        FileChannel source = null;
        FileChannel destination = null;
        source = new FileInputStream(sourceFile).getChannel();
        destination = new FileOutputStream(destFile).getChannel();
        if (destination != null && source != null) {
            destination.transferFrom(source, 0, source.size());
        }
        if (source != null) {
            source.close();
        }
        if (destination != null) {
            destination.close();
        }
    }

    public static Boolean verifyStoragePermissions(Activity activity) {
        if (Build.VERSION.SDK_INT >= 23) {
            int writePermission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE);
            int readPermission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.READ_EXTERNAL_STORAGE);

            if (writePermission != PackageManager.PERMISSION_GRANTED || readPermission != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(
                        activity,
                        PERMISSIONS_STORAGE,
                        REQUEST_EXTERNAL_STORAGE
                );
                return false;
            } else {
                return true;
            }
        } else {
            return true;
        }
    }

    public static Boolean verifyCameraPermissions(Activity activity) {
        if (Build.VERSION.SDK_INT >= 23) {
            int cameraPermission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.CAMERA);

            if (cameraPermission != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(
                        activity,
                        new String[]{(Manifest.permission.CAMERA)},
                        REQUEST_CAMERA
                );
                return false;
            } else {
                return true;
            }
        } else {
            return true;
        }
    }

    public static String getPath(final Context context, final Uri uri) {
        String result;
        Cursor cursor = context.getContentResolver().query(uri, null, null, null, null);
        if (cursor == null) {
            result = uri.getPath();
        } else {
            cursor.moveToFirst();
            int idx = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATA);
            result = cursor.getString(idx);
            cursor.close();
        }
        return result;
    }
}

File: croperino/src/main/java/com/mikelau/croperino/CroperinoFileUtil.kt
package com.mikelau.croperino

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Environment
import android.provider.MediaStore
import android.widget.Toast
import androidx.core.app.ActivityCompat
import com.mikelau.croperino.CroperinoConfig.Companion.getsDirectory
import com.mikelau.croperino.CroperinoConfig.Companion.getsImageName
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

object CroperinoFileUtil {
    var PERMISSIONS_STORAGE = arrayOf(
        Manifest.permission.READ_EXTERNAL_STORAGE,
        Manifest.permission.WRITE_EXTERNAL_STORAGE
    )
    const val REQUEST_EXTERNAL_STORAGE = 1
    const val REQUEST_CAMERA = 2
    var tempFile: File? = null
    fun setupDirectory(ctx: Context) {
        val state = Environment.getExternalStorageState()
        if (Environment.MEDIA_MOUNTED == state) {
            if (Environment.getExternalStorageDirectory().exists()) {
                tempFile = File(
                    Environment.getExternalStorageDirectory().toString() + getsDirectory(),
                    getsImageName()
                )
            } else {
                tempFile = File(
                    Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM)
                        .toString() + getsDirectory(), getsImageName()
                )
            }
        } else {
            tempFile = File(ctx.filesDir, getsImageName())
        }
        if (!tempFile!!.exists()) {
            tempFile = File(CroperinoConfig.getsRawDirectory())
            tempFile!!.mkdirs()
        }
    }

    @Throws(IOException::class)
    fun newCameraFile(): File? {
        val storageDir = File(
            Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES),
            "Camera"
        )
        if (!storageDir.exists()) {
            storageDir.mkdir()
        }
        tempFile = File.createTempFile(getsImageName(), ".jpg", storageDir)
        return tempFile
    }

    fun newGalleryFile(data: Intent, ctx: Activity): File? {
        return try {
            tempFile = File(CroperinoConfig.getsRawDirectory() + getsImageName())
            copyFile(File(getPath(ctx, data.data)), tempFile)
            tempFile
        } catch (e: Exception) {
            if (e is IOException) {
                tempFile = File(getPath(ctx, data.data))
            } else {
                Toast.makeText(ctx, "Gallery is empty or access is prohibited by device", Toast.LENGTH_SHORT).show()
            }
            tempFile
        }
    }

    @Throws(IOException::class)
    private fun copyFile(sourceFile: File, destFile: File?) {
        if (!sourceFile.exists()) {
            return
        }
        val source = FileInputStream(sourceFile).channel
        val destination = FileOutputStream(destFile).channel
        if (destination != null && source != null) {
            destination.transferFrom(source, 0, source.size())
        }
        source?.close()
        destination?.close()
    }

    fun verifyStoragePermissions(activity: Activity): Boolean {
        val writePermission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE)
        val readPermission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.READ_EXTERNAL_STORAGE)
        return if (writePermission != PackageManager.PERMISSION_GRANTED || readPermission != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(activity, PERMISSIONS_STORAGE, REQUEST_EXTERNAL_STORAGE)
            false
        } else {
            true
        }
    }

    fun verifyCameraPermissions(activity: Activity): Boolean {
        val cameraPermission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.CAMERA)
        return if (cameraPermission != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(activity, arrayOf(Manifest.permission.CAMERA), REQUEST_CAMERA)
            false
        } else {
            true
        }
    }

    fun getPath(context: Context, uri: Uri?): String? {
        val result: String?
        val cursor = context.contentResolver.query(uri!!, null, null, null, null)
        if (cursor == null) {
            result = uri.path
        } else {
            cursor.moveToFirst()
            val idx = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATA)
            result = cursor.getString(idx)
            cursor.close()
        }
        return result
    }
}
